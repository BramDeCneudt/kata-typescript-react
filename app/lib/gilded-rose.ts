import { StrategyService } from "./kata/strategy-service";

export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class GildedRose {
  items: Array<Item>;
  strategyService: StrategyService;

  constructor(items = [] as Array<Item>) {
    this.items = items;
    this.strategyService = new StrategyService
  }

  updateQuality() {

    for (const item of this.items) {
      const strategy = this.strategyService.generateStrategy(item);
      strategy.updateItem(item);
    }

    return this.items;
  }
}
