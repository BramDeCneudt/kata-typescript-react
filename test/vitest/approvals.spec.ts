import { execSync } from 'node:child_process';
import { Item, GildedRose } from '@/lib/gilded-rose';
import { describe, expect, test } from 'vitest';

/**
 * This test uses Vitest Snapshot, similar to [Jest Snapshot](https://goo.gl/fbAQLP).
 *
 * There are two test cases here with different styles:
 * <li>"foo" is more similar to the unit test from the 'Java' version
 * <li>"thirtyDays" is more similar to the TextTest from the 'Java' version
 *
 * I suggest choosing one style to develop and deleting the other.
 */

const items : Item[] = [
    new Item("+5 Dexterity Vest", 10, 0), //
    new Item("+6 Dexterity Vest", -1, 10), //
    new Item("Aged Brie", 2, 0), //
    new Item("Aged Brie", 2, 50), //
    new Item("Elixir of the Mongoose", 5, 7), //
    new Item("Sulfuras, Hand of Ragnaros", 0, 80), //
    new Item("Sulfuras, Hand of Ragnaros", -1, 80),
    new Item("Backstage passes to a TAFKAL80ETC concert", 15, 50),
    new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
    new Item("Backstage passes to a TAFKAL80ETC concert", 10, 50),
    new Item("Backstage passes to a TAFKAL80ETC concert", 10, 20),
    new Item("Backstage passes to a TAFKAL80ETC concert", 5, 20),
    new Item("Backstage passes to a TAFKAL80ETC concert", 5, 50),
    new Item("Backstage passes to a TAFKAL80ETC concert", 0, 49),
    new Item("Conjured Mana Cake", 3, 6),
    new Item("Conjured Mana Cake", -3, 6),
    new Item("Conjured Mana Cake", -3, 3)
]

describe('Gilded Rose Golden Test', () => {
  test('Golden Test base test', () => {
    const gildedRose : GildedRose = new GildedRose(items);
    const actualItems = gildedRose.updateQuality();

    expect(actualItems).toMatchSnapshot();
  });

  test('Golden Test test after 5 days', () => {
    const gildedRose : GildedRose = new GildedRose(items);
    let actualItems;
    for (let day = 0; day <= 5; day++) {
      actualItems = gildedRose.updateQuality() ;
    }

    expect(actualItems).toMatchSnapshot();
  });

});
