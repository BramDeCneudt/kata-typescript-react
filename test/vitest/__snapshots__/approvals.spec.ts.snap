// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gilded Rose Golden Test > Golden Test base test 1`] = `
[
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 0,
    "sellIn": 9,
  },
  Item {
    "name": "+6 Dexterity Vest",
    "quality": 8,
    "sellIn": -2,
  },
  Item {
    "name": "Aged Brie",
    "quality": 1,
    "sellIn": 1,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": 1,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 6,
    "sellIn": 4,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 14,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 21,
    "sellIn": 14,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 9,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 22,
    "sellIn": 9,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 23,
    "sellIn": 4,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 4,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -1,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 5,
    "sellIn": 2,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 4,
    "sellIn": -4,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 1,
    "sellIn": -4,
  },
]
`;

exports[`Gilded Rose Golden Test > Golden Test test after 5 days 1`] = `
[
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 0,
    "sellIn": 3,
  },
  Item {
    "name": "+6 Dexterity Vest",
    "quality": 0,
    "sellIn": -8,
  },
  Item {
    "name": "Aged Brie",
    "quality": 12,
    "sellIn": -5,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -5,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 0,
    "sellIn": -2,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 8,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 29,
    "sellIn": 8,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 3,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 36,
    "sellIn": 3,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -2,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -2,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -7,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 0,
    "sellIn": -4,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 0,
    "sellIn": -10,
  },
  Item {
    "name": "Conjured Mana Cake",
    "quality": 0,
    "sellIn": -10,
  },
]
`;
